### PR TITLE
[CB-5195][CB-4078]: correctOrientation not working on camera.getPicture for PHOTOLIBRARY

### PR DESCRIPTION
**Pull request by swbradshaw**
This fixes the orientation bug for Android 4.4+ devices when choosing a picture from the photo library.

This fixes both DATA_URL and FILE_URI requests.

Tested on Android 4.0.4,  Android 5.02, and Android 5.1 devices.

### CODE REVIEW

**Comment by kukikiloke** [comment]
@swbradshaw what devices did you see this issue?

I am seeing similar issue only on Samsung devices (S6 5.1.1 and S3 5.1.1) that selecting a photo taken in portrait from photo library does not correct the orientation (displayed in landscape)

**Comment by swbradshaw** [comment]
I personally saw the issue on on LG G2(5.0.2) and LG V10 (5.1).  

I believe it affects all devices running Android 4.4 and newer because the existing code won't correctly pull the orientation from the MediaStore on API19+

**Comment by riknoll** [comment]
@swbradshaw What gallery apps did you test against?

**Comment by swbradshaw** [comment]
@riknoll - Can you clarify? 

Are you referring to the apps that have implement the DocumentsProvider interface that show up on the unified UI picker on 4.4+ devices?   If so, I tested the default picker provided by Android. If you try to test another 3rd party one, it doesn't work. This doesn't work in the master branch, either. This is due to a bug with the "ouputModifiedBitmap"  function where it won't correctly generate a filename when the user picks a path that comes back from a 3rd party URI returned by Android's DocumentProvider. This push request isn't addressing that issue, but I can certainly fix that one as well if its been logged.  (The spelling error in that function name is driving me crazy, too).

OR are you referring to pre-4.4 devices that prompt the user to pick an app to choose an image from?  If this is the case, the code should handle these the same way it has done in the past.

**Comment by riknoll** [comment]
We support Google Photos and the stock Gallery app as far as I know. Have you tested recently? There were some changes to how we dereference URIs (though Google Drive still does not work). Mainly, I just wanted to check to make sure that you had verified this change with photos selected using both Google Photos and the stock Gallery app. I think the `ouputModifiedBitmap()` bug should be fixed in master (+1 on misspelling BTW) .

**Comment by swbradshaw** [comment]
`ouputModifiedBitmap()` is not fixed in master. `getRealPathFromURI_API19()` in FileHelper returns an empty string and thus the fileName cannot be generated for the temp file. I just tested this 5 minutes ago.

**Comment by swbradshaw** [comment]
Should I go ahead and fix `ouputModifiedBitmap()` for this pull request?  The reality is now that orientation is fixed, this is going to cause that function to get called more because the image will need to be rotated for output.

**Comment by riknoll** [comment]
If you have a fix in mind that would be great! I think there might already be a JIRA for that as well.

**Comment by swbradshaw** [comment]
@riknoll - I found the changes you were talking about. The changes in CB-10460 will make the `ouputModifiedBitmap()` function correctly because the FileHelper will now return null, which is what it was expecting.  There are no further changes on my side needed. With this change and CB-10460 combined, I can now pick photos from any Document Provider (Drive, Gallery, Photos, QuickPic) and it will successfully correct the orientation and resize the photo.

One suggestion I have is to rename CB-10460 to something like "Choosing pictures from 3rd party provider (Gallery, Photo, Drive) will result in error".   If the current title is published on the blog, users aren't going to know what "getRealPath return null in some cases" means. :-)

**Comment by riknoll** [comment]
Our release notes actually come from the commit messages themselves (so no going back, I'm afraid)

**Comment by unbeatableDeepak** [comment]
@swbradshaw I used your code but unable to get the rotated image from  drive, dropbox and
 as well as from the internal storage(see attachment), i tested on Samsung Galaxy S6 with Android v5.1.1.
Any solution for that?


screenshot_2016-03-31-12-34-32

**Comment by swbradshaw** [comment]
@unbeatableDeepak Grab the changes in CB-10460 as well and that should fix your issue. Please verify.

**Comment by unbeatableDeepak** [comment]
@swbradshaw can you please give me the link?

**Comment by swbradshaw** [comment]
@unbeatableDeepak please see this pull request:
https://github.com/apache/cordova-plugin-camera/pull/194

Can you please provide the code you are using? Internal storage should work.

**Comment by unbeatableDeepak** [comment]
[Archive.zip](https://github.com/apache/cordova-plugin-camera/files/197505/Archive.zip)
This is my code and its same as yours and CB-10460 changes is also there.

**Comment by swbradshaw** [comment]
@unbeatableDeepak Thanks. I was referring to your JS code. I would like to reproduce what you are seeing on my side with the same options you are calling to the plugin.

**Comment by swbradshaw** [comment]
@unbeatableDeepak I was able to reproduce your issues. I'll take a look to see how I can fix this.

**Comment by unbeatableDeepak** [comment]
@swbradshaw That's great.
When I debugged this issue I found that getRotationFromMediaStoreAPI19() is unable to get the orientation of the image for internal storage path. Also can you please check the issue for 3rd party apps?

**Comment by unbeatableDeepak** [comment]
@swbradshaw I forgot to give you my JS code, these are the options which I used:

navigator.camera.getPicture(function(imageUrl) {
    console.log("Image Data ::: "+imageUrl);
}, function(err) {
   console.log("Error ::: "+err);		
}, {
	quality : 50,
	destinationType : navigator.camera.DestinationType.FILE_URI,
	sourceType : navigator.camera.PictureSourceType.PHOTOLIBRARY,
	mediaType : navigator.camera.MediaType.PICTURE,
	correctOrientation : true
});

**Comment by unbeatableDeepak** [comment]
@swbradshaw Hi, did you get any solution yet?

**Comment by swbradshaw** [comment]
@unbeatableDeepak I have a solution coded, but I haven't had a chance to test it yet.

**Comment by unbeatableDeepak** [comment]
@swbradshaw That's great! If you don't have time then I can do the testing of your code and will let you know the result, can you please provide me your code so that I can test it.

**Comment by unbeatableDeepak** [comment]
@swbradshaw if your code is working please share it with me, my app's new release is pending due to this issue. 
Thanks in advance.

**Comment by swbradshaw** [comment]
@unbeatableDeepak My changes have been committed. I did a rebase from the master since I was weeks behind. Since this is my first PR, I'm not sure if that was correct or not. Anyway, you can test now.

**Comment by swbradshaw** [comment]
Re-opening with cleaned up fork.

**Comment by JEGardner** [comment]
This didn't compile for me.

See error:

:compileArmv7DebugJavaWithJavac/Users/jackg/Sites/bibendum/plonk/platforms/android/src/org/apache/cordova/camera/CameraLauncher.java:931: error: cannot find symbol
                            rotate = exifToDegrees(exif.getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_UNDEFINED));
                                     ^
  symbol:   method exifToDegrees(int)
  location: class CameraLauncher
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: /Users/jackg/Sites/bibendum/plonk/platforms/android/src/org/apache/cordova/file/AssetFilesystem.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
1 error
 FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileArmv7DebugJavaWithJavac'.

**Comment by swbradshaw** [comment]
@JEGardner Well, that's embarrassing. Check it out now and it should work for you.

**Comment by JEGardner** [comment]
@riknoll any chance we could get this merged so we can use the official repo and not a fork? Seems like a pretty important bug to fix?

**Comment by riknoll** [comment]
@JEGardner apologies, this is a pretty big change and I haven't had an opportunity to review/test it. I'll try and take a look at it this week. In the mean time, @swbradshaw there are a lot of whitespace changes here, can you try and remove those? They make the code tough to review.

**Comment by swbradshaw** [comment]
@riknoll I just removed a big chunk of the whitespace changes that were done by Android Studio. I find myself having to manually copy changes from my editor into a seperate text editor to commit, because there are so many indentation problems in this file. :-(  This is what led to the missing function earlier today.
I will need to do another pull request in the future to specifically to address indentation, spelling errors, and function documentation that is wrong. I was trying to fix so much. Sorry!

**Comment by cordova-qa** [comment]
Cordova CI Build has completed successfully. 

**Dashboard** - http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4/

**Results**
1. [Windows 8.1 Store]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=windows-slave,platformName=windows-8.1-store/) - [Console Output]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=windows-slave,platformName=windows-8.1-store/console) - [Test Results]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=windows-slave,platformName=windows-8.1-store/testReport/) - [Device Logs]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=windows-slave,platformName=windows-8.1-store/artifact/)
2. [Windows 10  Store]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=windows-slave,platformName=windows-10-store/) - [Console Output]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=windows-slave,platformName=windows-10-store/console) - [Test Results]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=windows-slave,platformName=windows-10-store/testReport/) - [Device Logs]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=windows-slave,platformName=windows-10-store/artifact/)
3. [Windows 8.1 Phone]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=windows-slave,platformName=windows-8.1-phone/) - [Console Output]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=windows-slave,platformName=windows-8.1-phone/console) - [Test Results]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=windows-slave,platformName=windows-8.1-phone/testReport/) - [Device Logs]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=windows-slave,platformName=windows-8.1-phone/artifact/)
4. [iOS]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=mac-slave,platformName=ios/) - [Console Output]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=mac-slave,platformName=ios/console) - [Test Results]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=mac-slave,platformName=ios/testReport/) - [Build Artifacts]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=mac-slave,platformName=ios/artifact/)
5. [Android Mac]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=mac-slave,platformName=android/) - [Console Output]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=mac-slave,platformName=android/console) - [Test Results]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=mac-slave,platformName=android/testReport/) - [Build Artifacts]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=mac-slave,platformName=android/artifact/)
6. [Android Windows]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=windows-slave,platformName=android/) - [Console Output]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=windows-slave,platformName=android/console) - [Test Results]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=windows-slave,platformName=android/testReport/) - [Device Logs]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/4//label=windows-slave,platformName=android/artifact/)

**Comment by swbradshaw** [comment]
@riknoll Have you had a chance to test these changes?

**Comment by riknoll** [comment]
@swbradshaw I apologize, I have not had a chance to test this out and I am not likely to anytime soon. Sorry, I should have left a comment

**Comment by stevengill** [comment]
Hey! Just reviewed this and it looks good to me. 

Can you do one final rebase from master and I'll merge it in?

**Comment by stevengill** [comment]
@swbradshaw any chance you can deal with the merge conflicts?

**Comment by swbradshaw** [comment]
@stevengill ill take a look this weekend. This is my first PR/rebase so I've got some homework to do

**Comment by stevengill** [comment]
Sounds good @swbradshaw. It looks like https://github.com/apache/cordova-plugin-camera/commit/fed798e6c7f56e23bff994ab2c11d316c8257a78 created some conflicts. I'll hold off on the release until I can merge this in. Let me know how it goes!

**Comment by swbradshaw** [comment]
@stevengill - I think this is good to go. My first PR/rebase, hopefully didn't messed it up!

-Rebased from master
-Verified rotation is working correctly after merge conflict resolved
-Tested fed987 / CB-11447 to make sure it works w/merge conflict resolved
-A side-effect of getting rotation/scaling working, is this PR allows the user to choose picture from any Document Provider now (including Google Drive), provided the appropriate calling options are set (scale || rotate || base64 output).

**Comment by cordova-qa** [comment]
Cordova CI Build has completed successfully.

**Commit**     - [Link](https://github.com/apache/cordova-plugin-camera/pull/197/commits/0ed6406864ff5a166fbc7fe226193e40c8528eec)
**Dashboard** - [Link](http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74/)

| Builder Name  | Console Output | Test Report | Device Logs  |
|     :---:     |     :---:      |   :---:     |     :---:    |
| [Windows 8.1 Store]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=windows-8.1-store/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=windows-8.1-store/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=windows-8.1-store/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=windows-8.1-store/artifact/) |
| [Windows 10  Store]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=windows-10-store/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=windows-10-store/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=windows-10-store/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=windows-10-store/artifact/) |
| [Windows 8.1 Phone]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=windows-8.1-phone/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=windows-8.1-phone/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=windows-8.1-phone/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=windows-8.1-phone/artifact/) |
| [iOS]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=ios/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=ios/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=ios/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=ios/artifact/) |
| [Android]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=android/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=android/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=android/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/74//PLATFORM=android/artifact/) |

**Comment by stevengill** [comment]
@swbradshaw looks you included a new file named `CameraLauncherMaster.java` in your commit by mistake

**Comment by swbradshaw** [comment]
@stevengill Man, I'm such an amateur.  Removed.

**Comment by stevengill** [comment]
@swbradshaw don't worry about it! We all make mistakes like that. 

Really appreciate the PR! I'll test and merge it in tomorrow. 

Thanks!

**Comment by cordova-qa** [comment]
Cordova CI Build has completed successfully.

**Commit**     - [Link](https://github.com/apache/cordova-plugin-camera/pull/197/commits/e77ffa6bc704a9b18d41ffeea18bbaf91d02e34c)
**Dashboard** - [Link](http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75/)

| Builder Name  | Console Output | Test Report | Device Logs  |
|     :---:     |     :---:      |   :---:     |     :---:    |
| [Windows 8.1 Store]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=windows-8.1-store/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=windows-8.1-store/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=windows-8.1-store/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=windows-8.1-store/artifact/) |
| [Windows 10  Store]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=windows-10-store/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=windows-10-store/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=windows-10-store/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=windows-10-store/artifact/) |
| [Windows 8.1 Phone]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=windows-8.1-phone/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=windows-8.1-phone/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=windows-8.1-phone/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=windows-8.1-phone/artifact/) |
| [iOS]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=ios/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=ios/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=ios/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=ios/artifact/) |
| [Android]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=android/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=android/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=android/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/75//PLATFORM=android/artifact/) |

File: src/android/CameraLauncher.java [hunk 1]
```diff
@@ -927,6 +859,21 @@ private void writeUncompressedImage(Uri src, Uri dest) throws FileNotFoundExcept
...
     }
+    /**
+     * In the special case where the default width, height and quality are unchanged
+     * we just write the file out to disk saving the expensive Bitmap.compress function.
+     *
+     * @param src
+     * @throws FileNotFoundException
+     * @throws IOException
+     */
+    private void writeUncompressedImage(Uri src, Uri dest) throws FileNotFoundException,
```
Comment (stevengill) [inline src/android/CameraLauncher.java]: I see two writeUncompressedImage functions. Is this correct? The second one is right above this definition.

Comment (stevengill) [inline src/android/CameraLauncher.java]: Actually, i guess you are just overloading.

**Comment by stevengill** [comment]
Merged! Thanks again @swbradshaw

**Comment by Bnaya** [comment]
wow finally!
Thank you all!

**Pull request closed without merging**